File: Python/Labs/Lab_8/ge.py
def ge_fw(matrix):
    
    # TODO for end cases
    v = len(matrix)
    if len(matrix) > len(matrix[0]):
        v = len(matrix[0])
        
    for k in range(0, v, 1):
        
        # Find 1st Row with non-zero and interchange to top
        for i in range (k, len(matrix), 1): # Go through all the rows
            if not matrix[i][k] == 0: # For the first non-zero col with
                x = list(matrix[k])
                matrix[k] = list(matrix[i])
                matrix[i] = list(x)
                break
        
        # Make all lower rows zero for the first col
        for i in range (k + 1, len(matrix), 1):
            if not matrix[i][k] == 0:
                x = matrix[i][k] / matrix[k][k]
                
                # Reduce the first col to 0
                for j in range (0, len(matrix[i]), 1):
                    matrix[i][j] -= x*matrix[k][j]

    # Remove -0s
    for i in range (0, len(matrix), 1):
        for j in range (0, len(matrix[i]), 1):
            if matrix[i][j] == -0:
                matrix[i][j] = 0.0
    
    return matrix

def ge_bw(matrix):
    
    v = len(matrix)
    if len(matrix) > len(matrix[0]):
        v = len(matrix[0])
    
    # Comment it throughly and do odd cases
    for k in range (v - 1, -1, -1):
        
        n = -1 # Remember the row
        for j in range (k, -1, -1): # For that row go through every col and find non-zero
             
            if not matrix[k][j] == 0:
                n = j
                x = matrix[k][j]
                
                for i in range (0, len(matrix[k]), 1):
                    matrix[k][i] = matrix[k][i] / x
                    
                break
            
        # IF there is a non-zero
        if not n == -1:
            
            # Go through all the rows for that col and check if they are nonzero if not make them
            for i in range (0, len(matrix), 1):
                
                if not i == k: # Don't Cancel the original row
                    
                    if not matrix[i][n] == 0:
                        
                        x = matrix[i][n]
                        
                        # Reduce the value in that col to 0
                        for j in range (0, len(matrix[i]), 1):
                            matrix[i][j] -= x*matrix[k][j]
                        
        test = False # Set False
        for i in range (0, len(matrix), 1):
            
            # See if any non-zero value, if there is make test True
            for j in range (k, len(matrix[i]), 1):
                if not matrix[i][j] == 0:
                    test = True
            
            # If test is False, entire row is 0
            if test == False:
                x = list(matrix[len(matrix) - 1])
                matrix[len(matrix) - 1] = list(matrix[i])
                matrix[i] = list(x)

    # Remove -0s
    for i in range (0, len(matrix), 1):
        for j in range (0, len(matrix[i]), 1):
            if matrix[i][j] == -0:
                matrix[i][j] = 0.0
    
    return matrix

File: C/Labs/Lab_8/ge.c
#include<stdio.h>
#include<stdlib.h>

int ge_fw(float *matrix, int rows, int cols, float *matrix_out);
int ge_bw(float *matrix, int rows, int cols, float *matrix_out);

int ge_fw(float *matrix, int rows, int cols, float *matrix_out) {

    if (matrix == NULL || rows == 0 || cols == 0 || matrix_out == NULL) {
        return -1;
    }

    int v = rows;
    if (rows > cols) {
        v = cols;
    }

    int k = 0;
    for (k = 0; k < v; k++) {

        // Interchange 1st Row
        int i = k;
        for (i = k; i < rows; i++) { // Go through all rows
            if (matrix[i*cols + k] != 0) { // For each non-zero col
                
                float x[cols];
                int j = 0;
                for (j = 0; j < cols; j++) {
                    x[j] = matrix[k*cols + j];
                }

                for (j = 0; j < cols; j++) {
                    matrix[k*cols + j] = matrix[i*cols + j];
                }

                for (j = 0; j < cols; j++) {
                    matrix[i*cols + j] = x[j];
                }
                
                break; // Once array is found stop
            }
        }
        float x = 0;
        i = k + 1;
        for (i = k + 1; i < rows; i++) {
            
            if (matrix[i*cols + k] != 0) {
                x = matrix[i*cols + k] / matrix[k*cols + k];
                
                int j = 0;
                for (j = 0; j < cols; j++) {
                    matrix[i*cols + j] -= x*matrix[k*cols + j];
                }
            }
        }
    }
    
    int i = 0, j = 0;
    for (i = 0; i < rows; i++) {
        for (j = 0; j < cols; j++) {
            if (matrix[i*cols + j] == -0.000) {
                matrix[i*cols + j] = 0.000;
            }
            matrix_out[i*cols + j] = matrix[i*cols + j];
        }
    }
    return 0;
}

int ge_bw(float *matrix, int rows, int cols, float *matrix_out) {

    if (matrix == NULL || rows == 0 || cols == 0 || matrix_out == NULL) {
        return -1;
    }

    int v = rows;
    if (rows > cols) {
        v = cols;
    }

    int k = v, n = -1, j = k, i = k;
    float x = 0;
    for (k = v - 1; k > -1; k--) {
        n = -1, j = k;
        
        for (j = k; j > -1; j--) {

            if (matrix[k*cols + j] != 0) {
                x = matrix[k*cols + j];
                n = j;
                i = 0;
                for (i = 0; i < cols; i++) {
                    matrix[k*cols + i] = matrix[k*cols + i] / x;
                }

                break;
            }
        }

        if (n != -1) {
            for (i = 0; i < rows; i++) {
                    if (i != k && matrix[i*cols + n] != 0) {
                        x = matrix[i*cols + n];
                        j = 0;
                        for (j = 0; j < cols; j++) {
                            matrix[i*cols + j] = matrix[i*cols + j] - x*matrix[k*cols + j];
    
                        }
                    }
                }
        }
    }

    i = 0, j = 0;
    for (i = 0; i < rows; i++) {
        for (j = 0; j < cols; j++) {
            if (matrix[i*cols + j] == -0.000) {
                matrix[i*cols + j] = 0.000;
            }
            matrix_out[i*cols + j] = matrix[i*cols + j];
        }
    }
    return 0;
}
